### PR TITLE
#3258 - CAS supplier errors - Part 2

### PR DESCRIPTION
**Pull request by guru-aot**
1. Captured the prevalidation errors when it happens during the CAS calls and saved it in the cas_suppliers table with supplierStatus - ManualIntervention.

> Given name required for creation.
> Address outside of Canada cannot be processed.

2. Captured the know errors during the CAS service calls and saved it in the cas_suppliers table with supplierStatus - ManualIntervention.
The know errors are split by "|" and saved as list in the error column in cas_suppliers table.

> createSupplierAndSite
> createSiteForExistingSupplier
> getSupplierInfoFromCAS

3. E2E test cases were created for the error capture for the above know error and existing E2E is changed to capture the prevalidation error.

### CODE REVIEW

File: sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-evaluation-result-processor.ts [hunk 1]
```diff
@@ -25,4 +28,45 @@ export abstract class CASEvaluationResultProcessor {
...
+    studentSupplier: StudentSupplierToProcess,
+    summary: ProcessSummary,
+    error: string[],
+    auditUserId: number,
+  ): Promise<ProcessorResult> {
+    summary.warn("A known error occurred during processing.");
+    const now = new Date();
+    const auditUser = { id: auditUserId } as User;
+    try {
+      await this.casSupplierRepo.update(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-evaluation-result-processor.ts:50]: Need to adjust the method to optionally receive the supplier-found data if a new site creation fails to comply with the below business AC.
```
Any of these errors should make the Supplier isValid = false and in cases where Supplier found but create Site failed, save any valid Supplier info
```

File: sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-evaluation-result-processor.ts [hunk 2]
```diff
@@ -25,4 +28,45 @@ export abstract class CASEvaluationResultProcessor {
...
+          supplierStatus: SupplierStatus.ManualIntervention,
+          supplierStatusUpdatedOn: now,
+          isValid: false,
+          updatedAt: now,
+          modifier: auditUser,
+          errors: error,
+        },
+      );
+      summary.info("Set supplier status to Manual Intervention due to error.");
+    } catch (updateError: unknown) {
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-evaluation-result-processor.ts:64]: Every single place we call it just `error` for simplicity. I would keep doing it.

File: sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-active-supplier-found-processor.ts [hunk 3]
```diff
@@ -61,8 +63,19 @@ export class CASActiveSupplierFoundProcessor extends CASEvaluationResultProcesso
       });
       summary.info("Created a new site on CAS.");
     } catch (error: unknown) {
-      summary.error("Error while creating a new site on CAS.", error);
-      return { isSupplierUpdated: false };
+      if (error instanceof CustomNamedError) {
+        if (error.name === CAS_BAD_REQUEST) {
+          summary.warn("Known CAS error while creating a new site on CAS.");
+          return await this.processBadRequestErrors(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-active-supplier-found-processor.ts:69]: No need to use `await`.

File: sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-known-errors-processor.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+@Injectable()
+export class CASKnownErrorsProcessor extends CASEvaluationResultProcessor {
+  constructor(
+    @InjectRepository(CASSupplier)
+    casSupplierRepo: Repository<CASSupplier>,
+    private readonly systemUsersService: SystemUsersService,
+  ) {
+    super(casSupplierRepo);
+  }
+  /**
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-known-errors-processor.ts:26]: Please add a blank line.

File: sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-known-errors-processor.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  async process(
+    studentSupplier: StudentSupplierToProcess,
+    evaluationResult: CASEvaluationResult,
+    summary: ProcessSummary,
+  ): Promise<ProcessorResult> {
+    if (evaluationResult.status !== CASEvaluationStatus.KnownErrors) {
+      throw new Error("Incorrect CAS evaluation result processor selected.");
+    }
+    summary.warn(
+      `CAS evaluation result status: ${evaluationResult.status}. Error: ${evaluationResult.knownErrors}.`,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-known-errors-processor.ts:43]: The log for `CAS evaluation result status:` is already happening from the service. Please follow the same from the other processors and also add a final success message, also following the other porcessors.

File: sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-supplier.models.ts [hunk 6]
```diff
@@ -42,6 +46,11 @@ export enum NotFoundReason {
   NoActiveSupplierFound = "Supplier found but not active",
 }
 
+export interface CASKnownErrorsResult {
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-supplier.models.ts:49]: Please add a comment.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cas-integration/_tests_/cas-supplier-integration.scheduler.e2e-spec.ts [hunk 7]
```diff
@@ -679,4 +687,166 @@ describe(describeProcessorRootTest(QueueNames.CASSupplierIntegration), () => {
...
+    });
+    expect(updateCASSupplier).toEqual({
+      id: savedCASSupplier.id,
+      isValid: false,
+      supplierStatus: SupplierStatus.ManualIntervention,
+      errors: ["0035: Site does not exist."],
+    });
+  });
+
+  it("Should insert a record in the casSuppliers table when createSiteForExistingSupplier throws a known error.", async () => {
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cas-integration/_tests_/cas-supplier-integration.scheduler.e2e-spec.ts:739]: This E2E will need to be adapted to as per https://github.com/bcgov/SIMS/pull/4017#discussion_r1870059940

File: sources/packages/backend/libs/integrations/src/cas/cas.service.ts [hunk 8]
```diff
@@ -267,6 +246,28 @@ export class CASService {
...
+      error instanceof AxiosError &&
+      error.response?.status === HttpStatus.BAD_REQUEST &&
+      !!error.response?.data[CAS_RETURNED_MESSAGES]
+    ) {
+      const casKnownErrors =
+        error.response.data[CAS_RETURNED_MESSAGES].toString();
+      throw new CustomNamedError(
+        "CAS Bad Request Errors",
+        CAS_BAD_REQUEST,
+        casKnownErrors.split("|"),
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/cas/cas.service.ts:265]: Was the API response checked to ensure that there are no white spaces between the "|"?

**Review by andrewsignori-aot — COMMENTED** [review]
Great work, please take a look at the comments.

File: sources/packages/backend/libs/integrations/src/cas/cas.service.ts (hunk 8 above)
Comment (guru-aot) [reply sources/packages/backend/libs/integrations/src/cas/cas.service.ts]: Nice one, i missed it, it was with spaces

`{
    "CAS-Returned-Messages": "[0034] SIN is already in use. | [9999] Duplicate Supplier , Reason: [0065]- Possible duplicate exists, please use online form"
}`

File: sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-evaluation-result-processor.ts [hunk 9]
```diff
@@ -25,4 +28,48 @@ export abstract class CASEvaluationResultProcessor {
...
+   * @param auditUserId audit user id.
+   * @param partialSupplier partial supplier information.ß
+   * @returns processor result.
+   */
+  async processBadRequestErrors(
+    studentSupplier: StudentSupplierToProcess,
+    summary: ProcessSummary,
+    error: string[],
+    auditUserId: number,
+    partialSupplier?: Partial<CASSupplier>,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-evaluation-result-processor.ts:46]: A type can be created in the models to be explicit about what the method should receive as shown below.
```ts
export type CASSupplierInfoForBadRequest = Pick<
  CASSupplier,
  "supplierNumber" | "supplierName" | "status" | "supplierProtected"
>;
```

File: sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-evaluation-result-processor.ts [hunk 10]
```diff
@@ -25,4 +28,48 @@ export abstract class CASEvaluationResultProcessor {
...
+          ...partialSupplier,
+          supplierStatus: SupplierStatus.ManualIntervention,
+          supplierStatusUpdatedOn: now,
+          isValid: false,
+          updatedAt: now,
+          modifier: auditUser,
+          errors: error,
+        },
+      );
+      summary.info("Set supplier status to Manual Intervention due to error.");
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-evaluation-result-processor.ts:66]: "Manual Intervention" in logs can be changed to use the enum `SupplierStatus.ManualIntervention`.

File: sources/packages/backend/libs/integrations/src/cas/cas.service.ts [hunk 11]
```diff
@@ -164,21 +165,10 @@ export class CASService {
...
-          error.response.data[CAS_RETURNED_MESSAGES],
-          CAS_BAD_REQUEST,
-        );
-      }
-      throw new Error("Error while creating supplier and site on CAS.", {
-        cause: error,
-      });
+      this.handleBadRequestError(
+        error,
+        "Error while creating supplier and site on CAS.",
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/cas/cas.service.ts:170]: The error message at lines 170 nd 211 are probably not intentionally the same. Can you check, please?

File: sources/packages/backend/libs/integrations/src/cas/cas.service.ts [hunk 12]
```diff
@@ -267,6 +246,28 @@ export class CASService {
...
+      error instanceof AxiosError &&
+      error.response?.status === HttpStatus.BAD_REQUEST &&
+      !!error.response?.data[CAS_RETURNED_MESSAGES]
+    ) {
+      const casKnownErrors =
+        error.response.data[CAS_RETURNED_MESSAGES].toString();
+      throw new CustomNamedError(
+        "CAS Bad Request Errors",
+        CAS_BAD_REQUEST,
+        casKnownErrors.split(" | "),
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/cas/cas.service.ts:265]: I would recommend executing the split by the "|" and then doing a `trim()` to remove spaces, if any.

File: sources/packages/backend/libs/test-utils/src/factories/cas-supplier.ts [hunk 13]
```diff
@@ -83,19 +83,22 @@ export function createFakeCASSupplier(
...
-  casSupplier.supplierName = `${faker.name.lastName()}, ${faker.name.firstName()}`;
-  casSupplier.supplierNumber = faker.datatype
-    .number({ min: 100000, max: 9999999999 })
-    .toString();
+  casSupplier.supplierName =
+    options?.initialValues?.supplierName ??
+    `${faker.name.lastName()}, ${faker.name.firstName()}`;
+  casSupplier.supplierNumber =
+    options?.initialValues?.supplierNumber ??
+    faker.datatype.number({ min: 100000, max: 9999999999 }).toString();
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/test-utils/src/factories/cas-supplier.ts:95]: These new properties may be added if the `if` condition is added to the expected status.
When the record is created they will not be set.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cas-integration/_tests_/cas-supplier-integration.scheduler.e2e-spec.ts [hunk 14]
```diff
@@ -679,4 +687,185 @@ describe(describeProcessorRootTest(QueueNames.CASSupplierIntegration), () => {
...
+      db,
+      { student },
+      {
+        initialValues: {
+          supplierName: "SMITH, MELANIE",
+          supplierNumber: "2006124",
+          supplierProtected: false,
+          status: "ACTIVE",
+        },
+      },
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/cas-integration/_tests_/cas-supplier-integration.scheduler.e2e-spec.ts:754]: I am not sure if I am following the E2E scenario. These data should come from the mocked CAS API result, right?

File: sources/packages/backend/libs/integrations/src/cas/cas.service.ts [hunk 15]
```diff
@@ -267,6 +246,28 @@ export class CASService {
     return supplierAddress;
   }
 
+  /**
+   * Handles bad request errors from CAS API.
+   * @param error The error object caught during the request.
+   * @param defaultMessage The default message to throw for non-specific errors.
+   */
+  private handleBadRequestError(error: unknown, defaultMessage: string): never {
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/cas/cas.service.ts:254]: In some of the upcoming PRs lets add a unit test to validate the `CustomNamedError`.

File: sources/packages/backend/libs/integrations/src/cas/cas.service.ts (hunk 8 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/integrations/src/cas/cas.service.ts]: As mentioned below, let's try to have a unit test to validate the error in some of the next PRs.
https://github.com/bcgov/SIMS/pull/4017#discussion_r1872276210

File: sources/packages/backend/libs/integrations/src/cas/cas.service.ts (hunk 12 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/integrations/src/cas/cas.service.ts]: The recommendation was to do the split by "|" not " | ".
We can review it when we add the unit test as below comment
https://github.com/bcgov/SIMS/pull/4017#discussion_r1872276210

**Review by andrewsignori-aot — APPROVED** [review]
Great work, looks good 👍 
In some of the upcoming PR lets ensure there will be at least one unit tests as mentioned here: 
https://github.com/bcgov/SIMS/pull/4017#discussion_r1872276210

File: sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-active-supplier-not-found-processor.ts [hunk 16]
```diff
@@ -65,8 +67,21 @@ export class CASActiveSupplierNotFoundProcessor extends CASEvaluationResultProce
       });
       summary.info("Created supplier and site on CAS.");
     } catch (error: unknown) {
-      summary.error("Error while creating supplier and site on CAS.", error);
-      return { isSupplierUpdated: false };
+      if (error instanceof CustomNamedError) {
+        if (error.name === CAS_BAD_REQUEST) {
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-active-supplier-not-found-processor.ts:71]: That could be one condition only:

```js
if (error instanceof CustomNamedError && error.name === CAS_BAD_REQUEST) {
```

File: sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-known-errors-processor.ts [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+  StudentSupplierToProcess,
+} from "../cas-supplier.models";
+import { CASEvaluationResultProcessor, ProcessorResult } from ".";
+import { InjectRepository } from "@nestjs/typeorm";
+import { CASSupplier } from "@sims/sims-db";
+import { Repository } from "typeorm";
+import { SystemUsersService } from "@sims/services";
+
+/**
+ * Assert the student can be added to CAS.
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-known-errors-processor.ts:15]: This description seems to be wrong. AI generated? 🤣

File: sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-known-errors-processor.ts [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+   */
+  async process(
+    studentSupplier: StudentSupplierToProcess,
+    evaluationResult: CASEvaluationResult,
+    summary: ProcessSummary,
+  ): Promise<ProcessorResult> {
+    if (evaluationResult.status !== CASEvaluationStatus.KnownErrors) {
+      throw new Error("Incorrect CAS evaluation result processor selected.");
+    }
+    summary.info("Known CAS Errors found.");
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-known-errors-processor.ts:43]: errors in lower case, please.

File: sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-known-errors-processor.ts (hunk 17 above)
Comment (guru-aot) [reply sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-known-errors-processor.ts:15]: No copy paste from a existing processor :)

File: sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-active-supplier-not-found-processor.ts (hunk 16 above)
Comment (guru-aot) [reply sources/packages/backend/apps/queue-consumers/src/services/cas-supplier/cas-evaluation-result-processor/cas-active-supplier-not-found-processor.ts:71]: For other CustomNamedError other than CAS_BAD_REQUEST we need to set the summary as error and isSupplierUpdated as false.

**Review by andrepestana-aot — APPROVED** [review]
Just minor comments. Looks good!

**Pull request merged**